Sources/SwiftFrameCore/Workers/ImageWriter.swift

```diff
@@ -0,0 +1,103 @@
+import AppKit
+import Foundation
+
+public final class ImageWriter {
+
+    // MARK: - Exporting
+
+    static func finish(
+        context: CGContext,
+        with outputPaths: [FileURL],
+        sliceSize: CGSize,
+        outputWholeImage: Bool,
+        locale: String,
+        suffix: String,
+        format: FileFormat,
+        completion: @escaping () throws -> Void) throws {
+        guard let image = context.makeImage() else {
+            throw NSError(description: "Could not render output image")
+        }
+        DispatchQueue.global().ky_asyncOrExit {
+            let slices = sliceImage(image, with: sliceSize)
+            var slicesFinished = false
+            var bigImageFinished = !outputWholeImage
+
+            let workGroup = DispatchGroup()
+            workGroup.enter()
+
+            // Writing images asynchronously gave a big performance boost, what a surprise
+            // Also, since we checked beforehand if the directory is writable, we can safely put of the rendering work to a different queue
+            DispatchQueue.global().ky_asyncOrExit {
+                try ImageWriter.write(images: slices, to: outputPaths, locale: locale, suffix: suffix, format: format)
+                slicesFinished = true
+                if slicesFinished && bigImageFinished {
```

Comment:
This is a bit odd. You're using a dispatch group but you only `enter` it once and then depending on which of the two operations finishes earlier, that one leaves the group. I think the easier solution would be to simply `enter` the group in both asynchronous operations and leave it at the end of each. That way `workGroup.wait` should wait until all operations are done.

Reply:
Ahhh I didn't know you could enter a dispatch group more than once and then also leave it more than once. Makes sense

Reply:
Just tried it out and it seems this does not work as intended. The goal was to not call the completion before writing every image to file. When calling `workGroup.enter()` and `workGroup.leave()` instead of these boolean checks, a lot of images don't get written at all since the program finishes too early